pyroomacoustics/doa/srp.py

```diff
@@ -63,24 +87,34 @@ def _process(self, X):
 
         CC = []
         for k in self.freq_bins:
-            CC.append( np.dot(pX[:,k,:], np.conj(pX[:,k,:]).T) )
+            CC.append(np.dot(pX[:, k, :], np.conj(pX[:, k, :]).T))
         CC = np.array(CC)
 
+        M = self.L.shape[1]
+        ar = np.arange(M)
+        av = ar[:, None]
+        mask_diag = av == av.T
+        mask_triu = av <= av.T
+
+        mask_triu2 = (av <= av.T).flatten()
+        mask_diag2 = mask_diag.flatten()[mask_triu2]
+        CC2 = CC.reshape((-1, CC.shape[-2] * CC.shape[-2]))[:, mask_triu2]
+
         for n in range(self.grid.n_points):
 
             # get the mode vector axis: (frequency, microphones)
-            mode_vec = self.mode_vec[self.freq_bins,:,n]
+            mode_vec = self.mode_vec[self.freq_bins, :, n]
 
             # compute the outer product along the microphone axis
-            mode_mat = np.conj(mode_vec[:,:,None]) * mode_vec[:,None,:]
-
-            # multiply covariance by mode vectors and sum over the frequencies
-            R = np.sum(CC * mode_mat, axis=0)
+            mode_mat = np.conj(mode_vec[:, :, None]) * mode_vec[:, None, :]
 
-            # Now sum over all distince microphone pairs
-            sum_val = np.inner(ones, np.dot(np.triu(R, 1), ones))
+            # method 2
```

Comment:
this is very elegant! But maybe a more descriptive comment to explain the logic (it took me some time to parse out). Something like: "sum real part of upper triangle of CC * mode_mat as matrix is hermitian.

Reply:
and do we need to add the diagonal components? as @Womac pointed out and on p. 80 of the thesis, the M autocorrelations are " independent of the steering delays". So from what I understand, this same offset should be present for all grid points. Which is why I think we used `np.triu(R, 1)` before, but the imaginary components were previously included, which was wrong.

Reply:
@Womac is indeed correct. However, keeping the offset may help when plotting the cost function. I think that in SRP-PHAT all the diagonal components should be one, and in that case, the constant offset is always the number of channels. I will check this again though. But in that case, we could just add this so that the cost function is always positive, which is not really necessary, but may be useful sometimes.

Reply:
@fakufaku, just an idea,  if we want to keep the diagonal, I think we can write this even shorter, couldn't we just `srp_f = np.real(np.conj(mode_vec[..., None, :])@ CC @ mode_vec[..., None])` where `np.real` is only there to avoid imaginary parts due to rounding errors. Afterwards, we could just sum over all frequencies to obtain the global estimate. If we store the grid_points in one of the first dimensions of mode_vec and put a dimension of one in CC there, we could also avoid the for-loop, then the shapes could be something like: `mode_vec.shape = (grid_points, frequencies, M)`, and `CC.shape = (1,frequencies, M, M)` In the end with `SRP = np.sum(srp_f, axis=1)` we get the SRP coefficients [not normalised yet], where `SRP.shape = (grid_points, 1,1)`? Finally, `spr_cost = SRP/ self.num_snap / self.num_freq / self.num_pairs` What do you think?

Reply:
@ebezzam I have added more comments. Is it easier to parse now ?

Reply:
@fakufaku yes really easy now!